docs/transaction.md

```diff
@@ -0,0 +1,261 @@
+
+```
+C1: enqueue msg to Shard1 to schedule T1
+C2: enqueue msg to Shard1 to schedule T2  # enqueued earlier than C1
+
+C1: enqueue msg to Shard2 to schedule T1
+C2: enqueue msg to Shard2 to schedule T2 # enqueued later than C1
+
+shard1: pull T2, add it to TxQueue, pull T1, add it to TxQueue
+shard2: pull T1, add it to TxQueue, pull T2, add it to TxQueue
+
+TxQueue1: T2, T1  # wrong order
+TxQueue2: T1, T2
+```
+
+
+Once the transaction is added to the tx-queue, the shard also marks the tx-keys using the *intent* locks. Those locks do not block the flow of the underlying operation but merely express the intent to touch or modify the key. In reality, they are represented by a map: `lock:str->counter`. If `lock[key] == 2` it means the tx-queue has 2 pending transactions that plan to modify `key`. These intent locks are used for optimizations detailed below and are not required to implement the naive version of VLL algorithm.
+
+Once the scheduling hops converges, it means that the transaction entered the execution phase, in which it never rollbacks, or retries. Once it's been scheduled, VLL guarantees the progress of subsequent execution operations while providing strict-serializability guarantees.
+
+It's important to note that a scheduled transaction does not hold exclusivity on its keys. There could be other transactions that still mutate the keys it touches - these transactions were scheduled earlier and have not finished running yet, or even have not even started running.
+
+## Executing a transaction
+
+Once the transaction is scheduled, the coordinator starts sending the execution messages. We break each command to one or more micro-ops and each operation corresponds to a single message hop.
+
+For example, "MSET" corresponds to a single micro-op "mset" that has the same semantics, but runs in parallel on all the involved shards.
+
+However, "RENAME" requires two micro-ops: fetching the data from two keys, and then the second hop - deleting/writing a key (depending whether the key is a source or a destination).
+
+When a coordinator schedules an execution, it also specifies whether this execution is the last hop for that command. This is necessary, so that shards could do clean-up operations when running the last execution request: unlocking the keys and removing the transaction from the tx-queue.
```

Comment:
dont use schedules, its confusing.
When a coordinator sends an execution message it also..

Reply:
I think you should switch the order of this paragraph and the next one

Reply:
replaced.